substrate/client/network/sync/src/strategy/chain_sync.rs

```diff
@@ -1045,6 +1026,28 @@ where
 		}
 	}
 
+	fn attempt_state_sync(
+		&mut self,
+		finalized_hash: B::Hash,
+		finalized_number: NumberFor<B>,
+		skip_proofs: bool,
+	) {
+		let mut heads: Vec<_> = self.peers.values().map(|peer| peer.best_number).collect();
+		heads.sort();
+		let median = heads[heads.len() / 2];
+		if finalized_number + STATE_SYNC_FINALITY_THRESHOLD.saturated_into() >= median {
+			if let Ok(Some(header)) = self.client.header(finalized_hash) {
+				log::debug!(
+					target: LOG_TARGET,
+					"Starting state sync for #{finalized_number} ({finalized_hash})",
+				);
+				self.state_sync =
+					Some(StateSync::new(self.client.clone(), header, None, None, skip_proofs));
+				self.allowed_requests.set_all();
+			}
```

Comment:
While you are at it, could you add an error logging if the header is not available? Also `debug_assert!(false)` would be useful.